### PR TITLE
Remove pow(floating, int) overloads

### LINKED ISSUE #890: <cmath>: pow(float, int) doesn't return double
This bug was reported a long time ago at Developer Community, but I couldn't find an open issue here. Has it been forgotten?

```
#include <cmath>
#include <iomanip>

int main()
{
    // This should be OK since C++11
    static_assert( std::is_same<decltype(std::pow(10.f, 2)), double>::value );
}
```

See also DevCom-183102 and Microsoft-internal VSO-553723.

### PR DESCRIPTION
**Pull request by Berrysoft**
Fixes #890

As MSVC supports at least C++14, I simply removed the pow(floating, int) overloads, because they should be removed after C++11.

I also found that the template one could be reduced to `_GENERIC_MATH2` macro, so I did it.

Tests are also added.

### CODE REVIEW

File: tests/tr1/tests/cmath1/test.cpp [hunk 1]
```diff
@@ -8,6 +8,8 @@
 #include "tdefs.h"
 #include <cmath>
 
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
```
Comment (cbezault) [inline tests/tr1/tests/cmath1/test.cpp:11]: I kind of wish there was a unified place `STATIC_ASSERT` could be defined. Not the end of the world though.

Comment (Berrysoft) [reply tests/tr1/tests/cmath1/test.cpp:11]: I searched for `STATIC_ASSERT` and found that it is defined in many source files, but all those previous definitions are in std tests.

File: tests/tr1/tests/cmath1/test.cpp [hunk 2]
```diff
@@ -50,6 +52,8 @@ void test_cpp() { // test C++ header
 
     double (*pabs)(double) = &STD abs; // C++ overloads
     CHECK((*pabs)(-rthalf) == rthalf);
+
+    STATIC_ASSERT(std::is_same<decltype(std::pow(10.0, 2)), double>::value);
```
Comment (StephanTLavavej) [inline tests/tr1/tests/cmath1/test.cpp:56]: Thanks for adding test coverage! However, as https://github.com/microsoft/STL/blob/master/tests/tr1/README.md explains, we generally don't add anything to `tr1` these days. (We'll update it for breaking changes.)

There doesn't seem to be a super great existing test in `tests/std` to add such coverage; `VSO_0000000_instantiate_containers` tests `pow` for `complex` so we could add separate tests for arithmetic `pow`. Alternatively, adding a new test directory would be reasonable.

In addition to these test cases which are exercising `(double, int)`, `(float, int)`, and `(long double, int)`, I'd like to see more extensive coverage (as the "sufficient additional overloads" wording applies to the first argument being integral too). For example, this is a comprehensive test for `complex` `pow`; the arithmetic cases would be strictly fewer:

https://github.com/microsoft/STL/blob/0cdf5fbfac83650bf6b404a754be415493466643/tests/std/tests/VSO_0000000_instantiate_containers/test.cpp#L866-L918

File: tests/std/tests/VSO_0099869_pow_float_overflow/test.cpp [hunk 3]
```diff
@@ -44,7 +44,7 @@ class test_std_pow_against_crt {
 
     void single(uint32_t baseCandidate) {
         float input   = reinterpret_as<float>(baseCandidate);
-        float powOut  = std::pow(input, 2);
+        float powOut  = input * input;
```
[the remark this answers is not in the document]
Comment (Berrysoft) [reply tests/std/tests/VSO_0099869_pow_float_overflow/test.cpp]: I'll try it.

File: tests/tr1/tests/cmath1/test.cpp (hunk 2 above)
Comment (Berrysoft) [reply tests/tr1/tests/cmath1/test.cpp]: Tests added.

**Comment by Berrysoft** [comment]
Seems that it has a internal name VSO-553723, so I use that as the name of test.

**Comment by miscco** [comment]
Could you please Tag the github issue of PR instead. MSFT internal Tickets are not publicly accessible.

**Comment by Berrysoft** [comment]
> @miscco Could you please Tag the github issue of PR instead.

Name changed. Waiting for checks.

**Comment by StephanTLavavej** [comment]
@CaseyCarter and @BillyONeal reminded me that we now have a `// COMPILE-ONLY` convention for tests that don't execute any code at runtime, so I pushed one more change to follow this convention.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this highly visible bug, simplifying the codebase, improving the test coverage, and congratulations on your first microsoft/STL commit! :tada: :smile_cat:

**Pull request merged**